pkg/service/blob_storage.go

```diff
@@ -0,0 +1,114 @@
+package service
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"mime"
+	"strings"
+	"time"
+
+	"github.com/gabriel-vasile/mimetype"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/instill-ai/pipeline-backend/pkg/recipe"
+	"github.com/instill-ai/pipeline-backend/pkg/resource"
+	"github.com/instill-ai/pipeline-backend/pkg/utils"
+
+	artifactpb "github.com/instill-ai/protogen-go/artifact/artifact/v1alpha"
+	resourcex "github.com/instill-ai/x/resource"
+)
+
+func (s *service) uploadBlobAndGetDownloadURL(ctx context.Context, ns resource.Namespace, data string) (string, error) {
+	mimeType, err := getMimeType(data)
+	if err != nil {
+		return "", fmt.Errorf("get mime type: %w", err)
+	}
+	artifactClient := s.artifactPublicServiceClient
+	requesterUID, _ := resourcex.GetRequesterUIDAndUserUID(ctx)
+
+	vars, err := recipe.GenerateSystemVariables(ctx, recipe.SystemVariables{})
+
+	if err != nil {
+		return "", fmt.Errorf("generate system variables: %w", err)
+	}
+
+	ctx = metadata.NewOutgoingContext(ctx, utils.GetRequestMetadata(vars))
+
+	timestamp := time.Now().Format(time.RFC3339)
+	objectName := fmt.Sprintf("%s-%s%s", requesterUID.String(), timestamp, getFileExtension(mimeType))
+
+	resp, err := artifactClient.GetObjectUploadURL(ctx, &artifactpb.GetObjectUploadURLRequest{
+		NamespaceId:      ns.NsID,
+		ObjectName:       objectName,
+		ObjectExpireDays: 0,
```

Comment:
If I understand correctly, we only use this to upload input data (e.g. an image) before continuing with the pipeline pre-trigger execution. Later, we upload the hole pipeline run input (all the input values, whether referenced through a URL or not) to Minio again. Why does this has no expiration and the pipeline run input doesn't?

Reply:
Oh, mainly, I copy my previous code from worker to here.
I will try to follow the way `uploadPipelineRunInputsToMinio` did.

If so, we may need to remove [artifact client from service layer](https://github.com/instill-ai/pipeline-backend/pull/904#discussion_r1869358566) as well.

Reply:
Now, I realise the previous design.

In previous design, we upload and read data by the `minio client` initialised by `NewMinioClientAndInitBucket`.
This design mainly focused on uploading data and reading data internally.
Specifically, it means the input/output are only known by pipeline-backend itself but not the external client like user or console.
e.g. `pipeline-runs/input/d3e72143-0140-4224-ac89-9818a622df29.json`

Under the current design, we upload and read data by the `minio client` initialised by `NewMinioClient`.
This design mainly focused on uploading data and reading data for external usage.
e.g. `http://localhost:8080/v1alpha/namespaces/admin/blob-urls/0e39466e-515d-4449-af5b-e78ca6b994f8`

So, for the Pipeline input data, we upload those by `uploadPipelineRunInputsToMinio`, which is from the previous design. Those data should be read internally. So, we don't have to expose the download URL

But, for the user input data, it should be read by external users / console. So, we have to expose the download URL.

As for the retention issue, it should be addressed in another ticket.
Improvement: add retention when upload blob data in VDP (ins-6857).

Further description:

- Pipeline input data means the JSON object contains the data and metadata.
- User input data means the data users upload.

Reply:
So, in summary, what we should is
- add the TODO in `ObjectExpireDays: 0,`, both `service` and `worker` pkg.
- write clearer document for this reason.

Reply:
> We upload Pipeline Input Data by `uploadPipelineRunInputsToMinio`, which does not expose the public URL. The URL will be used by pipeline run logger.

You mean that the response for the pipeline run list will return the data blob instead of the URL, right? And therefore it will need to fetch that info from MinIO while building the response.

Reply:
🤔 I think this is a leftover. In protobufs, `inputsReference` and `outputsReference`fields were [removed](https://github.com/instill-ai/protobufs/blob/main/vdp/pipeline/v1beta/pipeline.proto#L2067) from the `PipelineRun` message. I think their presence in `ComponentRun` is a leftover. I checked console and, though they're defined in the `PipelineRun` type (not in the component one), they aren't used.